File: examples/governance/operations_domain/lib/operations_domain/deployment/commands/decommission_deployment.rb
module OperationsDomain
  class Deployment
    module Commands
      class DecommissionDeployment
        include Hecks::Command
        emits "DecommissionedDeployment"

        attr_reader :deployment_id

        def initialize(deployment_id: nil)
          @deployment_id = deployment_id
        end

        def call
          existing = repository.find(deployment_id)
          if existing
            unless existing.status == "deployed"
              raise Hecks::Error, "Cannot DecommissionDeployment: status must be 'deployed', got '#{existing.status}'"
            end
            Deployment.new(
              id: existing.id,
              model_id: existing.model_id,
              environment: existing.environment,
              endpoint: existing.endpoint,
              purpose: existing.purpose,
              audience: existing.audience,
              deployed_at: existing.deployed_at,
              decommissioned_at: Time.now.to_s,
              status: "decommissioned"
            )
          else
            raise Hecks::Error, "Deployment not found: #{deployment_id}"
          end
        end
      end
    end
  end
end

File: examples/governance_static_go/operations_static_go/domain/decommission_deployment.go
package domain

import (
	"time"
	"fmt"
)

type DecommissionDeployment struct {
	DeploymentId string `json:"deployment_id"`
}

func (c DecommissionDeployment) CommandName() string { return "DecommissionDeployment" }

func (c DecommissionDeployment) Execute(repo DeploymentRepository) (*Deployment, *DecommissionedDeployment, error) {
	existing, err := repo.Find(c.DeploymentId)
	if err != nil {
		return nil, nil, err
	}
	if existing == nil {
		return nil, nil, fmt.Errorf("Deployment not found: %s", c.DeploymentId)
	}
	existing.UpdatedAt = time.Now()
	if err := existing.Validate(); err != nil {
		return nil, nil, err
	}
	if err := repo.Save(existing); err != nil {
		return nil, nil, err
	}
	event := DecommissionedDeployment{
		AggregateID: existing.ID,
		DeploymentId: c.DeploymentId,
		OccurredAt: time.Now(),
	}
	return existing, &event, nil
}
